# Monte Carlo 

A ideia principal do método Monte Carlo (MC) é resolver problemas considerando um espaço de números aleatórios.

Em muitos casos podemos trocar a situação de resolver um problema determinístico por um probabilístico análogo. 
Então a pergunta que surge é: Como resolver esse problema probabilístico? É aí que entra o MC. 
O método gera uma amostragem de números aleatórios que se repetem um conjunto de n vezes, em que n é o 
número de experimentos que deseja realizar buscando resolver o problema.

Poderíamos resumir o MC da seguinte maneira: é um método que gera números aleatórios para comporem amostras 
buscando resolver problemas probabilísticos.


## A constante $\pi$

Como primeiro exemplo de aplicação do método MC, vamos calcular o valor da constante $\pi$. De maneira determinística sabemos que o valor da constante $\pi$ surge da razão entre o perímetro de um círculo e o seu diâmetro. Para calcular o valor de $\pi$ usando MC, vamos considerar um círculo de raio 1cm inscrito em um quadrado cujo lado é igual a 2cm. A figura abaixo mostra o que estamos buscando representar.

<img src="circulo1.png" width="260" height="260">

Para propor o valor de $\pi$ vamos fazer a razão entre as áreas do círculo e a área do quadrado $\frac{A_c}{A_q}$. Com a intensão de facilitar nosso problema, vamos considerar somente o primeiro quadrante. Portanto vamos pegar um quarto da área do círculo, o lado do quadrado correspondente passa a ser 1cm. Logo nossa razão entre as áreas será: $$\frac{A_c/4}{A_q} = \frac{\pi r^2}{4 l^2}.$$

Como $l = 1$ (lado do quadrado) e $r = 1$ (raio do círculo), então temos que o valor de $\pi$ a partir da razão entre as áreas é:

$$\pi = 4\frac{A_c}{A_q}$$

Assim qualquer valor que esteja no intervalo de 0 até 1 corresponde a um ponto dentro do semi-círculo. Observe o ponto $P_1$ que esta dentro da área laranjada.

<img src="circulo2.png" width="260" height="260">

Portando se atribuimos valores aleatórios de x e y nos intervalos $0 \le x \le 1$ e $0 \le y \le 1$, poderemos estimar $\pi$ aplicando o vínculo $r = (x^2 + y^2)^{\frac{1}{2}}$. Uma vez que todos os pontos chutados estiverem na região que são compreendidas pelo raio estaremos estimando a área do círculo, e os pontos que não estiverem nesse intervalo estão relacionados ao número de vezes que realizamos o experimento. 

#### Como fazemos:





In [26]:
import numpy as np
import random


numero_de_experimentos = 100
area_do_circulo = 0
intervalo_inicial = 0
intervalo_final = 1

for exps in range(numero_de_experimentos):
    x = random.uniform(intervalo_inicial, intervalo_final)
    y = random.uniform(intervalo_inicial, intervalo_final)
    r = (x**2 + y**2) ** (1/2)
    if r <= 1:
        area_do_circulo += 1
    
pi = 4*(area_do_circulo/numero_de_experimentos)

erro_absoluto = np.abs(np.pi - pi)
erro_relativo = (erro_absoluto/np.pi) * 100

print(f'Valor de pi = {pi}')
print(f'Erro Relativo = {round(erro_relativo, 2)}%')

Valor de pi = 3.16
Erro Relativo = 0.59%


Vamos melhorar o código acima criando uma função para executar essa tarefa.

In [29]:
import numpy as np
import random


def calcula_pi(n_experimentos):
    '''
    Essa função recebe uma lista de valores com
    potências de 10 e retorna uma tupla de listas
    com os valores estimados de pi e o erro 
    relativo dos valores.
    '''
    
    interv_inicial = 0
    interv_final = 1
    
    valores_pi = []
    erros_relativos = []
    for exps in n_experimentos:
        area_do_circulo = 0        
        for exp in range(exps):
            x = random.uniform(interv_inicial, interv_final)
            y = random.uniform(interv_inicial, interv_final)
            r = (x**2 + y**2) ** (1/2)
            if r <= 1:
                area_do_circulo += 1
                
        pi = 4*(area_do_circulo/exp)
        
        erro_absoluto = np.abs(np.pi - pi)
        erro_relativo = (erro_absoluto/np.pi) * 100
        
        valores_pi.append(pi)
        erros_relativos.append(erro_relativo)
        
    return valores_pi, erros_relativos


numero_experimentos = [10, 10**2, 10**3, 10**4, 10**5, 10**6]
valores_pi, erros_relativos = calcula_pi(numero_experimentos)

print(valores_pi)
print(erros_relativos)

[3.5555555555555554, 3.393939393939394, 3.1631631631631634, 3.1383138313831385, 3.140591405914059, 3.1410831410831412]
[13.176848420903347, 8.032446219953204, 0.6866106447225853, 0.10436815234171279, 0.03187070337047298, 0.01621828680015768]


----

### Questões


1) Escreva um gráfico do logaritmo do número de experimentos pelo erro relativo.

2) Caso seja aumentado o número de experimentos a redução do erro relativo compensa o custo computacional empregado.


----

## A Integral

Utilizamos o método MC para expressar o valor da constante $\pi$. Agora vamos tentar resolver o caso de uma integral que conhecemos analiticamente seu resultado. 

Dado a seguinte integral:

$$\int_{1}^{7}\frac{1}{x} dx,$$

sabemos que a solução analítica dessa integral é 

$$\int_{1}^{7}\frac{1}{x} dx = ln(7) - ln(1) \approx 1,9459.$$

1) Neste primeiro instante vamos usar uma regra de aproximação como a regra do trapézio composta. 

Uma integral da forma $\int_{a=x_0}^{b=x_n}f(x)dx$ pode ser aproximada por um valor de $I_1$, em que $I_1$ é dado como:

$$I_1 = \frac{h}{2}(y_0 + y_1).$$

  Caso o intervalo de integração $[a, b]$ seja subdividido  em $n$ intervalos igualmente espaçados a cada dois pontos, então:
  
  $$I_1 = \frac{h}{2}(y_0 + y_1) + \frac{h}{2}(y_1+ y_2)+ \ldots +\frac{h}{2}(y_{n-1} + y_{n})$$
  
  $$I_1 = \frac{h}{2}(y_0 + 2y_1+ 2y_2+ \ldots +2y_{n-1} + y_{n})$$
  
  $$I_1 = \frac{h}{2}\sum_{h}^{n}c_iy_i,$$

com $c_0 = 1$, $c_n = 1$ e $c_i = 2$ para $i = 1, 2, 3, \ldots, n-1$. 

O código abaixo descreve a regra do trapézio composta para calcular o integral $\int_{1}^{7}\frac{1}{x} dx$.


In [43]:
import numpy as np


def funcao(x):
    '''
    Recebe uma lista de valores e retorna
    a função que deseja integrar.
    '''
    return 1/x

def regra_trapezio_comp(x, y):
    '''
    Recebe duas listas de valores e retorna
    o valor calculado pelo método da regra
    do trapézio composta.
    '''
    regra_trap = []
    n = len(x)
    for i in range(n):
        if i == 0 or i == n-1:
            regra_trap.append(y[i])
        else:
            regra_trap.append(2*y[i])
    I_1 = (h/2)*sum(regra_trap)

    return round(I_1, 4)


intervalo_ini = 1
intervalo_final = 7
sub_divisoes = 10
h = (intervalo_final - intervalo_ini) / sub_divisoes

x = [i for i in np.arange(intervalo_ini, intervalo_final+0.1, h)]
y = [round(funcao(i), 4) for i in x]
I_1 = regra_trapezio_comp(x, y)

print(f'Valor da Integral: I_1 = {I_1}')

Valor da Integral: I_1 = 1.9743


Observe como precisamos de apenas 10 subdivisões do intervalo $[1, 7]$ para obter um valor de 1,9742, cujo erro relativo em relação ao valor correto é de 1,4543 %.  

2) Agora podemos utilizar o método de Monte Carlo para calcular a mesma integral. A ideia de se resolver uma integral utilizando o MC, considera que poderemos tomar a integral $\int_{a=x_0}^{b=x_n}f(x)dx$, multiplicar e dividir por uma densidade de probabilidade $\rho(x)$, obtendo: 

$$I_2 = \int_{a=x_0}^{b=x_n}\frac{f(x)}{\rho(x)}\rho(x)dx.$$

Deste modo, para resolver a integral $I_2$ considera-se um valor aleatório $\eta$ em um intervalo $[a, b]$ e calculamos uma média $\big\langle \frac{f(\eta)}{\rho(\eta)} \big \rangle$, sendo a densidade de probabilidade dada por

$$\rho(\eta) = \frac{1}{b-a}.$$

Assim a solução da integral é

$$I_2 = (b-a) \langle f(\eta) \rangle_{n}$$

$$I_2 = (b-a) \frac{1}{n} \sum_{i}^{n} f(\eta_{i}) ,$$

em que n é a quantidade de números aleátorios $\eta$ que satisfazem a densidade de probabilidade. Vamos criar um programa para calcular a integral $I_2$ considerando diferentes valores de n, como uma lista de valores $[10, 10², 10³, 10⁴, 10⁵, 10⁶]$.

In [17]:
# Versão 1
# Programa que calcula valores de uma integral
# via Monte Carlo.


import numpy as np
import random


def funcao(x):
    '''
    Recebe um valor aleatório 
    e retorna a função que deseja integrar.
    '''
    return 1/x

intervalo_ini = 1
intervalo_final = 7
n = 10
f_eta = 0

for i in range(n):
    eta_i = random.uniform(intervalo_ini, intervalo_final)
    f_eta += funcao(eta_i)

I_2 = round((intervalo_final - intervalo_ini) * (f_eta/n), 4)    
    
print(f'Monte Carlo: I_2 = {I_2}')


Monte Carlo: I_2 = 1.8626


Note que no caso acima, com um conjunto de 10 números aleatórios gerados obtemos um valor de 1.8626, sendo o erro relativo em relação ao valor analítico de 4,28 %, ou seja um valor aceitável. Vale a pena ressaltar que se caso você executar novamente o código acima é possível obter um erro maior ou menor. Lembre-se o MC gera números aleatórios em cada lance. Portanto é necessário que nossa amostra seja maior para obtermos valores da integral com mais segurança.

In [1]:
# Versão 2
# Programa que calcula valores de uma integral
# via Monte Carlo.


import numpy as np
import random


def funcao(x):
    '''
    Recebe um valor aleatório 
    e retorna o valor da função 
    que deseja integrar.
    '''
    return 1/x

intervalo_ini = 1
intervalo_final = 7
n = [10, 10**2, 10**3, 10**5, 10**6]

for n_i in n:
    f_eta = 0
    for i in range(n_i):
        eta_i = random.uniform(intervalo_ini, intervalo_final)
        f_eta += funcao(eta_i)

    I_2 = round((intervalo_final - intervalo_ini) * (f_eta/n_i), 4)    
    print(f'Monte Carlo: I_2 = {I_2}')
    

Monte Carlo: I_2 = 1.9724
Monte Carlo: I_2 = 1.9209
Monte Carlo: I_2 = 2.0062
Monte Carlo: I_2 = 1.9457
Monte Carlo: I_2 = 1.9438


Se você executar o versão 2 do programa que calcula a integral via MC irá perceber que os valores da integral para as quantidade de números aleatórios de $10⁵$ e $10⁶$ não sofrem grandes modificações. Sendo o erro relativo para $10⁶$ correspondente a  0,056 %.

----

### Questões

1) Expresse o valor da integral $\int_{1}^{7}\frac{1}{x} dx$ utilizando a primeira regra de Simpson ou regra do 1/3. Justifique se essa regra é melhor ou pior que as utilizadas até aqui, ou seja, a regra do trapézio composta e o método de Monte Carlo.

2) Justifique porque no caso da integral $\int_{1}^{7}\frac{1}{x} dx$ a regra do trapézio composta pode expressar uma solução com uma quantidade de passos bem menor que o método de MC. Será que poderíamos afirmar que independentemente do tipo de integral a regra do trapézio sempre converge o valor da integral em um número de passos menos que o MC?

3) A partir da versão 2 do programa que calcula o valor da integral utilizando MC, escreva um gráfico do erro pela quantidade de passos $[10, 10², 10³, 10⁴, 10⁵, 10⁶]$.

----